lib/tapioca/compilers/dsl/base.rb

```diff
@@ -176,6 +176,15 @@ def compile_method_return_type_to_rbi(method_def)
           return_type
         end
 
+        sig { params(type: String).returns(String) }
+        def as_nilable_type(type)
+          if type.start_with?("T.nilable(", "::T.nilable(") || type == "T.untyped" || type == "::T.untyped"
+            type
+          else
+            "T.nilable(#{type})"
+          end
```

Comment:
I think it still possible to get a `T.nilable(T.untyped)` with `as_nilable("T.proc.returns(T.nilable(T.untyped))")`.

We should refactor the helper methods to all pass through a single point where we can check/assert/raise things. In case we generate incorrect RBIs we could stop/skip during the generation and explaining to the user how to report an issue instead of producing invalid RBIs and confused developers.

Reply:
Yes, we can, and that's why I want to build the types like how Sorbet runtime does it, instead of concatenating strings, in the long run. We should be able to do:
```ruby
T.nilable(T.proc.returns(T.nilable(T.untyped))).to_s
# => "T.nilable(T.proc.params().returns(T.nilable(T.untyped)))"
```
and that should always return the correct thing. (Not sure if anyone fixed Sorbet runtime to collapse `T.nilable(T.untyped)` to just `T.untyped` though)